pkg/apis/podsecuritypolicy/types.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podsecuritypolicy
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	// SysctlsPodSecurityPolicyAnnotationKey represents the key of a whitelist of
+	// allowed safe and unsafe sysctls in a pod spec. It's a comma-separated list of plain sysctl
+	// names or sysctl patterns (which end in *). The string "*" matches all sysctls.
+	SysctlsPodSecurityPolicyAnnotationKey string = "security.alpha.kubernetes.io/sysctls"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodSecurityPolicy governs the ability to make requests that affect the SecurityContext
+// that will be applied to a pod and container.
+type PodSecurityPolicy struct {
```

Comment:
I didn't expect a second internal type... I expected the approach we've taken with other objects as we've moved them out of the extensions group:
* define external type in destination API group
* define conversions from that external type to the existing internal type (see apps/v1/doc.go as an example)
* register external types as cohabitating resources for storage (this PR does this already)
* register existing internal type into new group's internal scheme (see apps/register.go#addKnownTypes as an example)

I would follow the approach ReplicaSet, DaemonSet, etc, took

Reply:
@liggitt Thank you for the guide! I followed all the steps:

> define external type in destination API group

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/staging/src/k8s.io/api/podsecuritypolicy/v1beta1/types.go

> define conversions from that external type to the existing internal type (see apps/v1/doc.go as an example)

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/pkg/apis/podsecuritypolicy/v1beta1/doc.go

> register external types as cohabitating resources for storage

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/cmd/kube-apiserver/app/server.go#L584

> register existing internal type into new group's internal scheme

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/pkg/apis/podsecuritypolicy/register.go#L47-L53

But my API wasn't registered and not available:
```console
$ head -2 restricted2.yaml
apiVersion: podsecuritypolicy.admission.k8s.io/v1beta1
kind: PodSecurityPolicy
$ kubectl create -f restricted.yaml 
error: unable to recognize "restricted.yaml": no matches for podsecuritypolicy.admission.k8s.io/, Kind=PodSecurityPolicy
$ kubectl api-versions | grep podsecurity
$ curl -sk https://localhost:6443/apis | grep podsecurity
$ curl -sk https://localhost:6443/apis/podsecuritypolicy.admission.k8s.io/v1beta1
404 page not found
```
I don't understand what I'm missing here :( Could you help me?